src/Parallel/Reduction.hpp

```diff
@@ -250,17 +253,17 @@ struct NoSection {};
  *
  * This function supports section reductions (see `Parallel::Section`). Pass
  * the `Parallel::Section` as the \p section argument, or pass
- * `Parallel::NoSection` to perform a reduction over the entire parallel
+ * `Parallel::no_section()` to perform a reduction over the entire parallel
  * component (default). Here's an example of a section reduction:
  *
  * \snippet Test_SectionReductions.cpp section_reduction
  */
 template <class Action, class SenderProxy, class TargetProxy, class... Ts,
           class SectionType = NoSection>
-void contribute_to_reduction(
-    ReductionData<Ts...> reduction_data, const SenderProxy& sender_component,
-    const TargetProxy& target_component,
-    const SectionType& section = NoSection{}) noexcept {
+void contribute_to_reduction(ReductionData<Ts...> reduction_data,
+                             const SenderProxy& sender_component,
+                             const TargetProxy& target_component,
+                             SectionType& section = no_section()) noexcept {
```

Comment:
consider passing by `not_null` pointer

Reply:
So you think that would be better here? Not sure about the pros and cons here

Reply:
I think it should be just as acceptable to pass a pointer to the static member as a non-const ref, and the pointer enforces the `make_not_null` at the call site indicating that the section is getting mutated; so, I'd be mildly in favor of the `not_null` even though it can't follow our ordering convention because defaults have to be at the end.

Reply:
Ok I looked at this a bit and figured the non-const ref works better here: The `section` is mainly taken as an argument to specify which elements to reduce over and mutating the cookie is a secondary effect. For return-by-reference functions I like the `make_not_null` syntax, but for mutable function arguments like this a non-const ref seems a better fit to me (it's recommended everywhere I looked).

Reply:
okay, seems like a reasonable distinction to make.

Reply:
As the person who originally proposed that we adopt the return-by-pointer policy, I strongly disagree with this.  The point of the policy is to make it clear to the caller that the argument can be mutated.  It has nothing to do with whether the argument is considered to be a return value or not (except in that return values are mutated).

(The ordering convention is unrelated to pointers-vs.-references.  Defaults are an acceptable reason to deviate from it.)

Reply:
Hmm, as I said above I do like the pattern for return-by-ref functions. However, all suggestions I find online when searching for 'pointer vs reference' are to generally prefer mutable references for function arguments unless pointer semantics are needed. I'm hesitant to go against the standard suggestions.. do you have compelling reasons to?

Reply:
Here's the style guide entry our rules are based on: https://drake.mit.edu/styleguide/cppguide.html#Reference_Arguments .  It is rather light on rationale ("references act weird"), but based on personal communication I can explain it better:

The rule is all about mutability.  It has nothing to do with return values, except that return values are, by necessity, mutable.  It allows you to reason about the state of your objects without having to look up the prototypes of every function you call.  If I write
```c++
T x{};
f(x);
```
then, in general, I do not know whether `x` has been modified without looking up `f`.  Following the style rule, I know it is unchanged, but would likely be different if I had written
```c++
T x{};
f(&x);
```
instead.

I actually can't think of an argument to apply the rule to return values but not "external state" or similar.  In fact, I think the second case is more important, because return values are usually pretty obvious.  If you have one I'd be interested in hearing it.

(And, in any case, you don't get to ignore the SpECTRE style guide without a good reason.  If you don't like the rule propose a change in an appropriate venue.)

Reply:
Ok, changed this in a fixup. Here's a few thoughts:

- We should make our style guide a lot clearer and possibly include some of your rationale. Currently it's mostly about return values. This is all it says about the issue:
    > Prefer return by value over pass-by-mutable-reference except when mutable reference provides superior performance (in practice if you need a mutable reference use const gsl::not_null<Type*> instead of Type&). The mutable references must be the first arguments passed to the function.
    > https://spectre-code.org/code_review_guide.html
- Our style guide also links to google, which actually says this:
    > ... while non-optional output and input/output parameters should usually be references (which cannot be null).
    > https://google.github.io/styleguide/cppguide.html#Inputs_and_Outputs
    Not sure actually how the link to another google style guide is related to this?

Reply:
Google changed their opinion on this very recently (maybe with in the last 12-18 months). From the wording I'm assuming they are now preferring references because references can't be null, but that's also the point of `gsl::not_null` and I've generally found it very valuable to know at the call site what I can expect to be modified.